lib/change_stream.js

```diff
@@ -365,13 +369,19 @@ function createChangeStreamCursor(self, options) {
     changeStreamStageOptions.allChangesForCluster = true;
   }
 
+  if (changeStreamStageOptions.startAfter && self.hasReceived) {
+    changeStreamStageOptions.resumeAfter = changeStreamStageOptions.startAfter;
+    delete changeStreamStageOptions.startAfter;
+  }
+
   const pipeline = [{ $changeStream: changeStreamStageOptions }].concat(self.pipeline);
   const cursorOptions = applyKnownOptions({}, options, CURSOR_OPTIONS);
   const changeStreamCursor = new ChangeStreamCursor(
     self.topology,
     new AggregateOperation(self.parent, pipeline, options),
     cursorOptions
   );
+  changeStreamCursor.once('init', () => (self.hasReceived = true));
```

Comment:
what's the goal here? I'm wary of using an event listener for this since they are so prone to leaks. Are you just trying to see if the cursor's initial `aggregate` command was run?

Reply:
Not quite whether it was run, but whether it has gotten a response back yet, updating the resumeToken to one not associated with an `invalidate` event. I can refactor it to not use events, this just seemed like a simple approach.

Reply:
I'd prefer not to use an event, especially with the internals of a cursor-like thing. There is an internal property [init](https://github.com/mongodb/node-mongodb-native/blob/master/lib/cursor/core_cursor.js#L544) which is set to `true` once the initial operation is run, I think you should be able to use that.

Reply:
There's another check [here](https://github.com/mongodb/node-mongodb-native/pull/2273/commits/31c26b49d26e7ed9921c65c7ecb2405a6f31fbe1#diff-30861ceb96e5170099331504d187951eR372) that I forgot to mention. This is to get prose test 18 to past:

```
    18. $changeStream stage for ChangeStream started with startAfter against a server >=4.1.1
    that has received at least one result
    - MUST include a resumeAfter option
    - MUST NOT include a startAfter option
```

In order to pass this, the ChangeStream needs to keep track of whether it has ever received a result *across* cursors as they are re-created upon resumable errors.

However, I've refactored to not use events, and added comments to clarify what's happening.